Modules/overlapped.c

```diff
@@ -1795,15 +1854,84 @@ _overlapped_Overlapped_WSARecvFrom_impl(OverlappedObject *self,
     self->read_from.address_length = sizeof(self->read_from.address);
 
     Py_BEGIN_ALLOW_THREADS
-    ret = WSARecvFrom((SOCKET)handle, &wsabuf, 1, &nread, &flags,
-                      (SOCKADDR*)&self->read_from.address,
-                      &self->read_from.address_length,
-                      &self->overlapped, NULL);
+        ret = WSARecvFrom((SOCKET)handle, &wsabuf, 1, &nread, &flags,
+            (SOCKADDR*)&self->read_from.address,
+            &self->read_from.address_length,
+            &self->overlapped, NULL);
     Py_END_ALLOW_THREADS
 
     self->error = err = (ret < 0 ? WSAGetLastError() : ERROR_SUCCESS);
+    switch (err) {
+    case ERROR_BROKEN_PIPE:
+        mark_as_completed(&self->overlapped);
+        return SetFromWindowsErr(err);
+    case ERROR_SUCCESS:
+    case ERROR_MORE_DATA:
+    case ERROR_IO_PENDING:
+        Py_RETURN_NONE;
+    default:
+        self->type = TYPE_NOT_STARTED;
+        return SetFromWindowsErr(err);
+    }
+}
 
-    switch(err) {
+
+/*[clinic input]
+_overlapped.Overlapped.WSARecvFromInto
+
+    handle: HANDLE
+    buf as bufobj: object
```

Comment:
```suggestion
    buf as bufobj: Py_buffer
```

`Py_buffer` converter can be used here.

Reply:
`WSARecvInto()` doesn't do that either, so wouldn't consistency be better here? Also I'm probably ignorant here but would argument clinic automatically generate the conversion code if I did that?

Reply:
> Also I'm probably ignorant here but would argument clinic automatically generate the conversion code if I did that?


Yes, Argument will generate the conversion code for buffer protocol.

Reply:
That doesn't seem to be true. I changed the signature and ran argument clinic but I don't see any conversion code being added.

Reply:
```diff
diff --git a/Modules/clinic/overlapped.c.h b/Modules/clinic/overlapped.c.h
index 4e96ee2bb65..16d6013ef2f 100644
--- a/Modules/clinic/overlapped.c.h
+++ b/Modules/clinic/overlapped.c.h
@@ -917,7 +917,7 @@ PyDoc_STRVAR(_overlapped_Overlapped_WSARecvFromInto__doc__,
 
 static PyObject *
 _overlapped_Overlapped_WSARecvFromInto_impl(OverlappedObject *self,
-                                            HANDLE handle, PyObject *bufobj,
+                                            HANDLE handle, Py_buffer *bufobj,
                                             DWORD size, DWORD flags);
 
 static PyObject *
@@ -925,17 +925,22 @@ _overlapped_Overlapped_WSARecvFromInto(OverlappedObject *self, PyObject *const *
 {
     PyObject *return_value = NULL;
     HANDLE handle;
-    PyObject *bufobj;
+    Py_buffer bufobj = {NULL, NULL};
     DWORD size;
     DWORD flags = 0;
 
-    if (!_PyArg_ParseStack(args, nargs, ""F_HANDLE"Ok|k:WSARecvFromInto",
+    if (!_PyArg_ParseStack(args, nargs, ""F_HANDLE"y*k|k:WSARecvFromInto",
         &handle, &bufobj, &size, &flags)) {
         goto exit;
     }
-    return_value = _overlapped_Overlapped_WSARecvFromInto_impl(self, handle, bufobj, size, flags);
+    return_value = _overlapped_Overlapped_WSARecvFromInto_impl(self, handle, &bufobj, size, flags);
 
 exit:
+    /* Cleanup for bufobj */
+    if (bufobj.obj) {
+       PyBuffer_Release(&bufobj);
+    }
+
     return return_value;
 }
-/*[clinic end generated code: output=9bc096b4db308930 input=a9049054013a1b77]*/
+/*[clinic end generated code: output=5c9b17890ef29d52 input=a9049054013a1b77]*/
diff --git a/Modules/overlapped.c b/Modules/overlapped.c
index 3ec765257e3..a6d0b7a7fb7 100644
--- a/Modules/overlapped.c
+++ b/Modules/overlapped.c
@@ -1880,7 +1880,7 @@ _overlapped_Overlapped_WSARecvFrom_impl(OverlappedObject *self,
 _overlapped.Overlapped.WSARecvFromInto
 
     handle: HANDLE
-    buf as bufobj: object
+    buf as bufobj: Py_buffer
     size: DWORD
     flags: DWORD = 0
     /
@@ -1890,9 +1890,9 @@ Start overlapped receive.
 
 static PyObject *
 _overlapped_Overlapped_WSARecvFromInto_impl(OverlappedObject *self,
-                                            HANDLE handle, PyObject *bufobj,
+                                            HANDLE handle, Py_buffer *bufobj,
                                             DWORD size, DWORD flags)
-/*[clinic end generated code: output=45fc5d883a11c4e5 input=8eacd80d50157434]*/
+/*[clinic end generated code: output=30c7ea171a691757 input=4be4b08d03531e76]*/
 {
     if (self->type != TYPE_NONE) {
         PyErr_SetString(PyExc_ValueError, "operation already attempted");

```

It works.

Reply:
@agronholm could you try to implement @kumaraditya303 suggestion?

Reply:
Yup, it took me a while to understand that the generated code was added in `overlapped.c.h` and not `overlapped.c`.

Reply:
Ok, done. Note that the resulting code differs quite a bit from the existing functions using `Py_buffer`.

Reply:
Thanks.
I think that existing functions should be updated as well. In a separate PR.